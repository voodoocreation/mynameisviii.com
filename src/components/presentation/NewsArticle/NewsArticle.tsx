import cn from "classnames";
import * as React from "react";
import { MdAccessTime } from "react-icons/md";
import {
  FormattedMessage,
  injectIntl,
  WrappedComponentProps
} from "react-intl";

import { INewsArticle } from "../../../models/root.models";
import Schema from "../../schema/NewsArticle";
import ButtonBar from "../ButtonBar/ButtonBar";
import DateTime from "../DateTime/DateTime";
import Image from "../Image/Image";
import Link from "../Link/Link";
import NewsArticleType from "../NewsArticleType/NewsArticleType";
import PageHeader from "../PageHeader/PageHeader";

import "./NewsArticle.scss";

interface IProps extends INewsArticle, WrappedComponentProps {}

interface IState {
  isImageLoaded: boolean;
}

class NewsArticle extends React.Component<IProps, IState> {
  public readonly state: IState = {
    isImageLoaded: false
  };

  public render() {
    const { intl, ...article } = this.props;

    return (
      <article
        className={cn("NewsArticle", { isRendered: this.state.isImageLoaded })}
      >
        <PageHeader>{article.title}</PageHeader>

        <div className="NewsArticle--body">
          <Image
            alt={article.title}
            className="NewsArticle--image"
            src={article.imageUrl}
            onLoad={this.onImageLoad}
          />

          <section className="NewsArticle--meta">
            <div className="NewsArticle--posted">
              <MdAccessTime />{" "}
              <FormattedMessage
                id="POSTED_ON_DATE_BY_AUTHOR"
                values={{
                  author: article.author,
                  date: <DateTime value={article.createdAt} />
                }}
              />
            </div>

            <NewsArticleType value={article.type} />
          </section>

          <section className="NewsArticle--content">
            <div
              className="NewsArticle-content--text"
              dangerouslySetInnerHTML={{ __html: article.content }}
            />

            <ButtonBar className="NewsArticle--action">
              {this.renderAction()}
            </ButtonBar>
          </section>
        </div>

        <Schema {...article} />
      </article>
    );
  }

  private renderAction = () => {
    const { action } = this.props;

    if (action && action.route) {
      return (
        <Link className="Button isStyled" route={action.route} prefetch>
          {action.text}
        </Link>
      );
    }

    if (action && action.url) {
      return (
        <Link className="Button isStyled" href={action.url} isExternal>
          {action.text}
        </Link>
      );
    }

    return null;
  };

  private onImageLoad = () => {
    this.setState({
      isImageLoaded: true
    });
  };
}

export default injectIntl(NewsArticle);
