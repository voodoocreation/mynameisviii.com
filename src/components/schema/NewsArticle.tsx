import * as React from "react";

import { absUrl } from "../../helpers/dataTransformers";
import { INewsArticle, organization } from "../../models/root.models";
import Schema from "./Schema";

const publisher = organization({
  email: "mgmt@mynameisviii.com",
  logo:
    "https://s3.amazonaws.com/mynameisviii-static/voodoo-creation-records-logo.png",
  name: "Voodoo Creation Records"
});

const NewsArticle: React.FC<INewsArticle> = props => (
  <Schema
    {...{
      "@id": absUrl(`/news/${props.slug}`),
      "@type": "NewsArticle",
      articleBody: props.content,
      author: {
        "@type": "Person",
        name: props.author
      },
      dateModified: props.createdAt,
      datePublished: props.createdAt,
      description: props.excerpt,
      headline: props.title,
      image: props.imageUrl,
      mainEntityOfPage: {
        "@id": absUrl(`/news/${props.slug}`)
      },
      name: props.title,
      publisher: {
        "@type": publisher.type,
        email: publisher.email,
        logo: {
          "@type": "ImageObject",
          url: publisher.logo
        }
      },
      text: props.content,
      thumbnailUrl: props.ogImageUrl
    }}
  />
);

export default NewsArticle;
