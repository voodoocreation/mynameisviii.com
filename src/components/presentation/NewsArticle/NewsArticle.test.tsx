import { mount, render } from "enzyme";
import * as React from "react";

import NewsArticle from "./NewsArticle";

const g: any = global;

const setup = (fn: any, fromTestProps?: any) => {
  const props = {
    author: "author",
    content: "<p>content</p>",
    createdAt: "2017-10-10T18:00:00",
    excerpt: "excerpt",
    imageUrl: "imageUrl",
    isActive: true,
    slug: "slug",
    title: "title",
    ...fromTestProps
  };

  return {
    actual: fn(<NewsArticle {...props} />),
    props
  };
};

describe("[presentation] <NewsArticle />", () => {
  beforeEach(() => {
    Object.defineProperty(g.Image.prototype, "complete", {
      value: false
    });
  });

  it("renders correctly", () => {
    const { actual } = setup(render);
    expect(actual).toMatchSnapshot();
  });

  it("updates `isImageLoaded` state when image has loaded", () => {
    const { actual } = setup(mount);

    actual.find(".NewsArticle-image img").simulate("load");
    expect(actual.render()).toMatchSnapshot();
  });

  it("updates `isImageLoaded` state when image has previously been loaded", () => {
    Object.defineProperty(g.Image.prototype, "complete", {
      value: true
    });

    const { actual } = setup(mount);

    actual.find(".NewsArticle-image img").simulate("load");
    expect(actual.render()).toMatchSnapshot();
  });
});
