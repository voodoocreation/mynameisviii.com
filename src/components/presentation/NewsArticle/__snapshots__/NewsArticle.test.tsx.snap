// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[presentation] <NewsArticle /> when no action is defined matches snapshot 1`] = `
<article
  class="NewsArticle"
>
  <header
    class="PageHeader"
  >
    <h1>
      Title
    </h1>
  </header>
  <div
    class="NewsArticle--body"
  >
    <figure
      class="Image NewsArticle--image isLoading"
    >
      <div
        class="Image--wrapper"
      >
        <img
          alt="Title"
          src="Image URL"
        />
        <span
          class="Loader"
        >
          <div
            class="Loader--animation"
          >
            <span />
            <span />
            <span />
          </div>
        </span>
      </div>
    </figure>
    <section
      class="NewsArticle--meta"
    >
      <div
        class="NewsArticle--posted"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8zm.5-13H11v6l5.25 3.15.75-1.23-4.5-2.67z"
          />
        </svg>
         Posted 
        <time
          class="DateTime"
          datetime="2017-10-10T18:00:00"
          title="Tuesday, 10 October 2017, 6:00 pm"
        >
          last year
        </time>
         by Author
      </div>
      <div
        class="NewsArticleType NewsArticleType-news"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M23 12l-2.44-2.78.34-3.68-3.61-.82-1.89-3.18L12 3 8.6 1.54 6.71 4.72l-3.61.81.34 3.68L1 12l2.44 2.78-.34 3.69 3.61.82 1.89 3.18L12 21l3.4 1.46 1.89-3.18 3.61-.82-.34-3.68L23 12zm-10 5h-2v-2h2v2zm0-4h-2V7h2v6z"
          />
        </svg>
         
        <span
          class="NewsArticleType--label"
        >
          News
        </span>
      </div>
    </section>
    <section
      class="NewsArticle--content"
    >
      <div
        class="NewsArticle-content--text"
      >
        <p>
          Content
        </p>
      </div>
    </section>
  </div>
  <script
    type="application/ld+json"
  >
    {"@id":"http://localhost:/news/article-1","@type":"NewsArticle","articleBody":"&lt;p&gt;Content&lt;/p&gt;","author":{"@type":"Person","name":"Author"},"dateModified":"2017-10-10T18:00:00","datePublished":"2017-10-10T18:00:00","description":"Excerpt","headline":"Title","image":"Image URL","mainEntityOfPage":{"@id":"http://localhost:/news/article-1"},"name":"Title","publisher":{"@type":"LocalBusiness","email":"mgmt@mynameisviii.com","logo":{"@type":"ImageObject","url":"https://s3.amazonaws.com/mynameisviii-static/voodoo-creation-records-logo.png"}},"text":"&lt;p&gt;Content&lt;/p&gt;","thumbnailUrl":"","@context":"http://schema.org/"}
  </script>
</article>
`;

exports[`[presentation] <NewsArticle /> when the action is defined, with a route matches snapshot 1`] = `
<article
  class="NewsArticle"
>
  <header
    class="PageHeader"
  >
    <h1>
      Title
    </h1>
  </header>
  <div
    class="NewsArticle--body"
  >
    <figure
      class="Image NewsArticle--image isLoading"
    >
      <div
        class="Image--wrapper"
      >
        <img
          alt="Title"
          src="Image URL"
        />
        <span
          class="Loader"
        >
          <div
            class="Loader--animation"
          >
            <span />
            <span />
            <span />
          </div>
        </span>
      </div>
    </figure>
    <section
      class="NewsArticle--meta"
    >
      <div
        class="NewsArticle--posted"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8zm.5-13H11v6l5.25 3.15.75-1.23-4.5-2.67z"
          />
        </svg>
         Posted 
        <time
          class="DateTime"
          datetime="2017-10-10T18:00:00"
          title="Tuesday, 10 October 2017, 6:00 pm"
        >
          last year
        </time>
         by Author
      </div>
      <div
        class="NewsArticleType NewsArticleType-news"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M23 12l-2.44-2.78.34-3.68-3.61-.82-1.89-3.18L12 3 8.6 1.54 6.71 4.72l-3.61.81.34 3.68L1 12l2.44 2.78-.34 3.69 3.61.82 1.89 3.18L12 21l3.4 1.46 1.89-3.18 3.61-.82-.34-3.68L23 12zm-10 5h-2v-2h2v2zm0-4h-2V7h2v6z"
          />
        </svg>
         
        <span
          class="NewsArticleType--label"
        >
          News
        </span>
      </div>
    </section>
    <section
      class="NewsArticle--content"
    >
      <div
        class="NewsArticle-content--text"
      >
        <p>
          Content
        </p>
      </div>
      <div
        class="ButtonBar NewsArticle--action"
      >
        <a
          class="Button isStyled"
          href="/"
        >
          Action
        </a>
      </div>
    </section>
  </div>
  <script
    type="application/ld+json"
  >
    {"@id":"http://localhost:/news/article-1","@type":"NewsArticle","articleBody":"&lt;p&gt;Content&lt;/p&gt;","author":{"@type":"Person","name":"Author"},"dateModified":"2017-10-10T18:00:00","datePublished":"2017-10-10T18:00:00","description":"Excerpt","headline":"Title","image":"Image URL","mainEntityOfPage":{"@id":"http://localhost:/news/article-1"},"name":"Title","publisher":{"@type":"LocalBusiness","email":"mgmt@mynameisviii.com","logo":{"@type":"ImageObject","url":"https://s3.amazonaws.com/mynameisviii-static/voodoo-creation-records-logo.png"}},"text":"&lt;p&gt;Content&lt;/p&gt;","thumbnailUrl":"","@context":"http://schema.org/"}
  </script>
</article>
`;

exports[`[presentation] <NewsArticle /> when the action is defined, with a url matches snapshot 1`] = `
<article
  class="NewsArticle"
>
  <header
    class="PageHeader"
  >
    <h1>
      Title
    </h1>
  </header>
  <div
    class="NewsArticle--body"
  >
    <figure
      class="Image NewsArticle--image isLoading"
    >
      <div
        class="Image--wrapper"
      >
        <img
          alt="Title"
          src="Image URL"
        />
        <span
          class="Loader"
        >
          <div
            class="Loader--animation"
          >
            <span />
            <span />
            <span />
          </div>
        </span>
      </div>
    </figure>
    <section
      class="NewsArticle--meta"
    >
      <div
        class="NewsArticle--posted"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8zm.5-13H11v6l5.25 3.15.75-1.23-4.5-2.67z"
          />
        </svg>
         Posted 
        <time
          class="DateTime"
          datetime="2017-10-10T18:00:00"
          title="Tuesday, 10 October 2017, 6:00 pm"
        >
          last year
        </time>
         by Author
      </div>
      <div
        class="NewsArticleType NewsArticleType-news"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M23 12l-2.44-2.78.34-3.68-3.61-.82-1.89-3.18L12 3 8.6 1.54 6.71 4.72l-3.61.81.34 3.68L1 12l2.44 2.78-.34 3.69 3.61.82 1.89 3.18L12 21l3.4 1.46 1.89-3.18 3.61-.82-.34-3.68L23 12zm-10 5h-2v-2h2v2zm0-4h-2V7h2v6z"
          />
        </svg>
         
        <span
          class="NewsArticleType--label"
        >
          News
        </span>
      </div>
    </section>
    <section
      class="NewsArticle--content"
    >
      <div
        class="NewsArticle-content--text"
      >
        <p>
          Content
        </p>
      </div>
      <div
        class="ButtonBar NewsArticle--action"
      >
        <a
          class="Button isStyled"
          href="/"
          rel="noopener"
          target="_blank"
        >
          Action
        </a>
      </div>
    </section>
  </div>
  <script
    type="application/ld+json"
  >
    {"@id":"http://localhost:/news/article-1","@type":"NewsArticle","articleBody":"&lt;p&gt;Content&lt;/p&gt;","author":{"@type":"Person","name":"Author"},"dateModified":"2017-10-10T18:00:00","datePublished":"2017-10-10T18:00:00","description":"Excerpt","headline":"Title","image":"Image URL","mainEntityOfPage":{"@id":"http://localhost:/news/article-1"},"name":"Title","publisher":{"@type":"LocalBusiness","email":"mgmt@mynameisviii.com","logo":{"@type":"ImageObject","url":"https://s3.amazonaws.com/mynameisviii-static/voodoo-creation-records-logo.png"}},"text":"&lt;p&gt;Content&lt;/p&gt;","thumbnailUrl":"","@context":"http://schema.org/"}
  </script>
</article>
`;
